crates/shared/src/zeroex_api.rs

```diff
@@ -183,24 +373,80 @@ pub enum ZeroExResponseError {
 #[async_trait::async_trait]
 impl ZeroExApi for DefaultZeroExApi {
     async fn get_swap(&self, query: SwapQuery) -> Result<SwapResponse, ZeroExResponseError> {
-        self.swap_like_request_impl("quote", query).await
+        self.request(query.format_url(&self.base_url, "quote"))
+            .await
     }
 
     async fn get_price(&self, query: SwapQuery) -> Result<PriceResponse, ZeroExResponseError> {
-        self.swap_like_request_impl("price", query).await
+        self.request(query.format_url(&self.base_url, "price"))
+            .await
+    }
+
+    async fn get_orders_with_pagination(
+        &self,
+        query: &OrdersQuery,
+        results_per_page: usize,
+        page: usize,
+    ) -> Result<OrdersResponse, ZeroExResponseError> {
+        let mut url = query.format_url(&self.base_url);
+        url.query_pairs_mut()
+            .append_pair("page", &page.to_string())
+            .append_pair("perPage", &results_per_page.to_string());
+        self.request(url).await
+    }
+
+    async fn get_orders(
+        &self,
+        query: &OrdersQuery,
+    ) -> Result<Vec<OrderRecord>, ZeroExResponseError> {
+        let mut results = Vec::default();
+        let mut page = 1;
+        loop {
+            let response = self
+                .get_orders_with_pagination(query, ORDERS_MAX_PAGE_SIZE, page)
+                .await?;
+            if !expect_more_results_after_handling_response(&mut results, response) {
+                break;
+            }
+            page += 1;
+        }
+        retain_valid_orders(&mut results);
+        Ok(results)
     }
 }
 
+/// Append data of response to results and return whether another page should be fetched.
+fn expect_more_results_after_handling_response(
+    results: &mut Vec<OrderRecord>,
+    mut response: OrdersResponse,
+) -> bool {
+    // only expect another page if this one was full
+    let expect_more_results = response.records.len() as u64 == response.per_page;
+    results.append(&mut response.records);
+    expect_more_results
+}
+
+fn retain_valid_orders(orders: &mut Vec<OrderRecord>) {
+    let mut included_orders = HashSet::new();
+    let now = chrono::offset::Utc::now();
+    orders.retain(|order| {
+        // only keep orders which are still valid
+        order.order.expiry > now
+            // properly signed
+            && Signature::try_from(order.order.signature.clone()).is_ok()
+            // not a duplicate
+            && included_orders.insert(order.order.salt.clone())
```

Comment:
```suggestion
            && included_orders.insert(order.metadata.hash)
```

I don't think the salt is necessarily unique (it looks like a Unix timestamp to me). The order hash is though.

Reply:
Yeah that makes way more sense. I misread and thought the `salt` itself would be unique and not that the `salt` was added to make 2 otherwise identical orders unique.